src/transformers/models/doge/modular_doge.py

```diff
@@ -0,0 +1,1172 @@
+
+        # calculate dynamic mask from value_states
+        dt_states = self.dt_proj(
+            value_states.transpose(1, 2).reshape(value_states.shape[0], value_states.shape[-2], -1)
+        )
+        dynamic_mask = torch.exp(self.A * F.softplus(dt_states)).transpose(-1, -2)
+        attn_mask = self.prepare_dynamic_mask(
+            hidden_states=hidden_states,
+            dynamic_mask=dynamic_mask,
+            dynamic_mask_ratio=self.dynamic_mask_ratio,
+            attention_mask=attention_mask,
+        )
+
+        attention_interface: Callable = self.eager_attention_forward
+        if self.config._attn_implementation != "eager":
+            attention_interface = self.ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
+        attn_output = attention_interface(
+            query_states,
+            key_states,
+            value_states,
+            attention_mask=attn_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            scaling=self.scaling,
+            **kwargs,
+        )
+
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        return attn_output
+
+    def prepare_dynamic_mask(
+        self,
+        hidden_states: torch.Tensor,
+        dynamic_mask: torch.Tensor,
+        dynamic_mask_ratio: float = 0.0,
+        attention_mask: Optional[torch.Tensor] = None,
+    ):
+        """
+        Combine `dynamic_mask` with `attention_mask` to generate the final `attn_mask`.
+
+        Args:
+            hidden_states (`torch.Tensor`): The input hidden_states, used to determine the minimum value of the current input precision.
+            dynamic_mask (`torch.Tensor`): dynamic mask of shape `(batch_size, num_heads, key_sequence_length)`.
+            dynamic_mask_ratio (`float`, *optional*): Ratio from 0.0 to 1.0 used to control the proportion of the dynamic mask filled with the minimum value.
+            attention_mask (`torch.Tensor`, *optional*): attention mask of shape `(batch_size, 1, query_sequence_length, key_sequence_length)`.
+        """
+        attn_mask = None
+        if dynamic_mask is not None:
+            attn_mask = dynamic_mask[:, :, None, :]
+            if 0.0 < dynamic_mask_ratio < 1.0:
+                min_type = torch.finfo(hidden_states.dtype).min
+                num_dynamic_mask = int(attn_mask.shape[-1] * dynamic_mask_ratio)
+                if num_dynamic_mask > 0:
+                    rate_value = torch.kthvalue(attn_mask, num_dynamic_mask, dim=-1, keepdim=True).values
+                    attn_mask = attn_mask.masked_fill(attn_mask < rate_value, min_type)
+            if attention_mask is not None:
+                attn_mask = attn_mask + attention_mask[:, :, :, : attn_mask.shape[-1]]
+        else:
+            attn_mask = attention_mask
+
+        return attn_mask
+
+    def eager_attention_forward(
```

Comment:
the three attention forward look general (similar to Llama and the ones in `integration/flex` and `integrations/sdpa`. And as my comment above mentioned, you won't have flash if you use `self.ATTENTION` instead of the global scoped one!

Reply:
(Making it incompatible with TGI / vllm 😉 )